### ISSUE TITLE
How to use CSTM_label method.

### ISSUE THREAD

**Issue by jay25208** [YvanYin/Metric3D#43]
感谢作者团队的出色的工作！
我想复现训练，但对pytorch不太熟悉，可以提供些建议么？
此外我有点小问题，请指教下哈。
1、公布的推理代码是只用了变化深度标签CSTM_label方法吗？没用CSTM_image吧？
[pred_depth = pred_depth * normalize_scale / scale_info](https://github.com/YvanYin/Metric3D/blob/377e6c6642d0aca7aaa5a19e58fdcf5d0fd3d910/mono/utils/do_test.py#L166)

所以如果我训练，只要按照CSTM_label方法，把深度图数据按焦距系数转换就像吗？训练分辨率是怎么设置的，和crop_size的关系是什么，网络可以输入不同分辨率的数据训练吗？

2、如果是CSTM_label，实际上比例是 ，
[label_scale_factor = cano_label_scale_ratio * [resize_label_scale_ratio](https://github.com/YvanYin/Metric3D/blob/377e6c6642d0aca7aaa5a19e58fdcf5d0fd3d910/mono/utils/do_test.py#L208)，比论文中的cano_label_scale_ratio](url) 还多了个 resize_label_scale_ratio。而resize_label_scale_ratio是因为输入图像要调整成crop_size造成的，调整是先resize再pad。这个label_scale_factor 可以补充解释下么？
3、内参有作为模型的输入，在推理或者训练中使用吗？
项目中有相机内存模型构建方式，应该是构建了cam_model 内容是  [[x_center, y_center, fov_x, fov_y]](https://github.com/YvanYin/Metric3D/blob/377e6c6642d0aca7aaa5a19e58fdcf5d0fd3d910/mono/utils/transform.py#L391C26-L391C60)  ：
```
cam_model = build_camera_model(reshape_h, reshape_w, intrinsic)，
cam_model = np.stack([x_center, y_center, fov_x, fov_y], axis=2)
```
且[data字典](https://github.com/YvanYin/Metric3D/blob/377e6c6642d0aca7aaa5a19e58fdcf5d0fd3d910/mono/utils/do_test.py#L151C1-L154C6)的构建有input图像和cam_model，
```
    data = dict(
        input=input,
        cam_model=cam_model,
    )
```

但实际推理中好像又没用到cam_model ？
这个cam_model 是什么作用？有参与推理或者训练吗？

谢谢！

**Comment by JUGGHM** [comment]
> 感谢作者团队的出色的工作！ 我想复现训练，但对pytorch不太熟悉，可以提供些建议么？ 此外我有点小问题，请指教下哈。 1、公布的推理代码是只用了变化深度标签CSTM_label方法吗？没用CSTM_image吧？ [pred_depth = pred_depth * normalize_scale / scale_info](https://github.com/YvanYin/Metric3D/blob/377e6c6642d0aca7aaa5a19e58fdcf5d0fd3d910/mono/utils/do_test.py#L166)
> 
> 所以如果我训练，只要按照CSTM_label方法，把深度图数据按焦距系数转换就像吗？训练分辨率是怎么设置的，和crop_size的关系是什么，网络可以输入不同分辨率的数据训练吗？
> 
> 2、如果是CSTM_label，实际上比例是 ， [label_scale_factor = cano_label_scale_ratio * [resize_label_scale_ratio](https://github.com/YvanYin/Metric3D/blob/377e6c6642d0aca7aaa5a19e58fdcf5d0fd3d910/mono/utils/do_test.py#L208)，比论文中的cano_label_scale_ratio](url) 还多了个 resize_label_scale_ratio。而resize_label_scale_ratio是因为输入图像要调整成crop_size造成的，调整是先resize再pad。这个label_scale_factor 可以补充解释下么？ 3、内参有作为模型的输入，在推理或者训练中使用吗？ 项目中有相机内存模型构建方式，应该是构建了cam_model 内容是 [[x_center, y_center, fov_x, fov_y]](https://github.com/YvanYin/Metric3D/blob/377e6c6642d0aca7aaa5a19e58fdcf5d0fd3d910/mono/utils/transform.py#L391C26-L391C60) ：
> 
> ```
> cam_model = build_camera_model(reshape_h, reshape_w, intrinsic)，
> cam_model = np.stack([x_center, y_center, fov_x, fov_y], axis=2)
> ```
> 
> 且[data字典](https://github.com/YvanYin/Metric3D/blob/377e6c6642d0aca7aaa5a19e58fdcf5d0fd3d910/mono/utils/do_test.py#L151C1-L154C6)的构建有input图像和cam_model，
> 
> ```
>     data = dict(
>         input=input,
>         cam_model=cam_model,
>     )
> ```
> 
> 但实际推理中好像又没用到cam_model ？ 这个cam_model 是什么作用？有参与推理或者训练吗？
> 
> 谢谢！

感谢关注我们的工作，关于您的问题解答如下：
1、
（1）是的，CSTM_label相比于CSTM_image，允许图像自由地resize
（2）事实上分两步，将焦距对齐到公共空间，这一步变换depth和焦距；随机resize，这一步变换depth、内参中的光心和图像坐标（显然图像大小变了，坐标范围也变了）
（3）训练分辨率即crop_size，在随机resize后crop

2、参考1（2），第一步cano，第二步resize，pad适用于resize后、图像大小仍然小于crop_size的情况

3、cam_model在CSTM_label方案下已弃用

未来我们应该会开源训练代码揭露更多细节

**Comment by Zador-Pataki** [comment]
Hi, thanks for the great work. I am trying to understand this thread with google translate. Did I understand correctly that this repository implements CSTM_label and not CSTM_image? In section 4.2 of the paper you mention you implement CSTM_image on a number of applications. If this repo implements CSTM_label, can you share some insight on why you selected *_label here but chose *_image in the paper? 

Thanks!

**Comment by YvanYin** [comment]
@Zador-Pataki Yes. We find that CSTM_label has a little better performance, thus our released code use CSTM_label. Our released training code will use CSTM_label as well.
